loopout_test.go

```diff
@@ -99,3 +102,163 @@ func TestLateHtlcPublish(t *testing.T) {
+	)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	// Set up the required dependencies to execute the swap.
+	sweeper := &sweep.Sweeper{Lnd: &lnd.LndServices}
+	blockEpochChan := make(chan interface{})
+	statusChan := make(chan SwapInfo)
+	expiryChan := make(chan time.Time)
+	timerFactory := func(expiry time.Duration) <-chan time.Time {
+		return expiryChan
+	}
+
+	errChan := make(chan error)
+	go func() {
+		err := swap.execute(context.Background(), &executeConfig{
+			statusChan:     statusChan,
+			blockEpochChan: blockEpochChan,
+			timerFactory:   timerFactory,
+			sweeper:        sweeper,
+		}, ctx.Lnd.Height)
+		if err != nil {
+			logger.Error(err)
+		}
+		errChan <- err
+	}()
+
+	// The swap should be found in its initial state.
+	cfg.store.(*storeMock).assertLoopOutStored()
+	state := <-statusChan
+	if state.State != loopdb.StateInitiated {
+		t.Fatal("unexpected state")
+	}
+
+	// We'll then pay both the swap and prepay invoice, which should trigger
+	// the server to publish the on-chain HTLC.
+	signalSwapPaymentResult := ctx.AssertPaid(swapInvoiceDesc)
+	signalPrepaymentResult := ctx.AssertPaid(prepayInvoiceDesc)
+
+	signalSwapPaymentResult(nil)
+	signalPrepaymentResult(nil)
+
+	// Notify the confirmation notification for the HTLC.
+	ctx.AssertRegisterConf()
+
+	blockEpochChan <- int32(ctx.Lnd.Height + 1)
+
+	htlcTx := wire.NewMsgTx(2)
+	htlcTx.AddTxOut(&wire.TxOut{
+		Value:    int64(swap.AmountRequested),
+		PkScript: swap.htlc.PkScript,
+	})
+
+	ctx.NotifyConf(htlcTx)
+
+	// The client should then register for a spend of the HTLC and attempt
+	// to sweep it using the custom confirmation target.
+	ctx.AssertRegisterSpendNtfn(swap.htlc.PkScript)
+
+	expiryChan <- time.Now()
+
+	cfg.store.(*storeMock).assertLoopOutState(loopdb.StatePreimageRevealed)
+	status := <-statusChan
+	if status.State != loopdb.StatePreimageRevealed {
+		t.Fatalf("expected state %v, got %v",
+			loopdb.StatePreimageRevealed, status.State)
+	}
+
+	// assertSweepTx performs some sanity checks on a sweep transaction to
+	// ensure it was constructed correctly.
+	assertSweepTx := func(expConfTarget int32) *wire.MsgTx {
+		t.Helper()
+
+		sweepTx := ctx.ReceiveTx()
+		if sweepTx.TxIn[0].PreviousOutPoint.Hash != htlcTx.TxHash() {
+			t.Fatalf("expected sweep tx to spend %v, got %v",
+				htlcTx.TxHash(), sweepTx.TxIn[0].PreviousOutPoint)
+		}
+
+		// The fee used for the sweep transaction is an estimate based
+		// on the maximum witness size, so we should expect to see a
+		// lower fee when using the actual witness size of the
+		// transaction.
+		fee := btcutil.Amount(
+			htlcTx.TxOut[0].Value - sweepTx.TxOut[0].Value,
+		)
+
+		weight := blockchain.GetTransactionWeight(btcutil.NewTx(sweepTx))
+		feeRate, err := ctx.Lnd.WalletKit.EstimateFee(
+			context.Background(), expConfTarget,
+		)
+		if err != nil {
+			t.Fatalf("unable to retrieve fee estimate: %v", err)
+		}
+		minFee := feeRate.FeeForWeight(weight)
+
+		if minFee > fee {
```

Comment:
Why not equal here?

Reply:
Sorry, missed that. Maybe you could add some margin, to at least verify that it actually publishes with the lower fee rate.

Reply:
Wouldn't adding a margin be the same as getting the weight estimate as done in `GetSweepFee`? That's why I opted for it initially, as we really care about the confirmation target being used for the sake of the test, not the fee itself.